File: 23/sol2.py
import time
import sys
import re
import itertools

class Nanobot:
    def __init__(self, x, y, z, r):
        self.x = x
        self.y = y
        self.z = z
        self.r = r
        
        self.hash = x
        self.hash = self.hash*33 + y
        self.hash = self.hash*33 + z
        self.hash = self.hash*33 + r
    
    def intersects_with(self, other):
        dist = abs(self.x - other.x)\
            + abs(self.y - other.y)\
            + abs(self.z - other.z)
        return dist <= self.r + other.r
    
    def __contains__(self, point):
        dist = abs(self.x - point.x)\
            + abs(self.y - point.y)\
            + abs(self.z - point.z)
        return dist <= self.r
    
    def __eq__(self, other):
        if self == other:
            return True
        if type(self) != type(other):
            return False
        return self.x == other.x\
            and self.y == other.y\
            and self.z == other.z\
            and self.r == other.r
    
    def __hash__(self):
        return self.hash

    def __str__(self):
        return f"pos=<{self.x},{self.y},{self.z}>, r={self.r}"

class NanobotCollection:
    def __init__(self, nanobots):
        self.nanobots = set(nanobots)

    def intersects_with(self, other_nanobot):
        for my_nanobot in self.nanobots:
            if not my_nanobot.intersects_with(other_nanobot):
                return False
        return True

    def __len__(self):
        return len(self.nanobots)

    def add(self, nanobot):
        self.nanobots.add(nanobot)

    def __contains__(self, nanobot):
        return nanobot in self.nanobots
    
    def __iter__(self):
        return iter(self.nanobots)
    
    def __str__(self):
        return "{" + ", ".join(map(str, self.nanobots)) + "}"

class Point:
    def __init__(self, x, y, z):
        self.x = x
        self.y = y
        self.z = z
        
    def __str__(self):
        return f"Point({self.x}, {self.y}, {self.z})"

t1 = time.time()

#pos=<0,0,0>, r=4
LINE_RE = re.compile(r'^pos=<(.*),(.*),(.*)>, r=(.*)$')

nanobots = []
for line in sys.stdin:
    nanobot_params = map(int, re.findall(LINE_RE, line)[0])
    nanobot = Nanobot(*nanobot_params)
    nanobots.append(nanobot)

nanobot_collections = [NanobotCollection([n]) for n in nanobots]
for nanobot_collection in nanobot_collections:
    for nanobot in nanobots:
        if nanobot in nanobot_collection:
            continue
        if nanobot_collection.intersects_with(nanobot):
            nanobot_collection.add(nanobot)

max_collection = max(nanobot_collections, key=len)
print(len(max_collection))
"""
        Ovaj postupak pretpostavlja da je tražena točka jedan od vrhova manhattan sfera.
        To ne mora biti slučaj (npr. ako pričamo o 2D: sjecište dva pravokutnika; pravokutnik je u općem slučaju sjecište dva kvadrata)
"""
for n in max_collection:
    deltas = [-1, 0, 1]
    for dx, dy, dz in itertools.product(deltas, deltas, deltas):
        p = Point(n.x + dx*n.r, n.y + dy*n.r, n.z + dz*n.r)
        for other_n in max_collection:
            if p not in other_n:
                break
        else:
            print(p)

print("time passed: ", time.time() - t1)

File: year2018/23/sol2.cpp
#include <cstdio>
#include <iostream>
#include <vector>
#include <cmath>
#include <set>
using namespace std;

typedef long long ll;

inline int dist(const auto& n1, const auto& n2) {
    return abs(n1.x-n2.x) + abs(n1.y-n2.y) + abs(n1.z-n2.z);
}

struct point {
    int x, y, z;
    point(int x, int y, int z) : x(x), y(y), z(z) {}
};

struct nanobot {
    int x, y, z, r;
    nanobot(int x, int y, int z, int r) : x(x), y(y), z(z), r(r) {}
    
    bool contains(const point &p) const {
        return dist(*this, p) <= this->r;
    }
};

bool operator<(const nanobot& n1, const nanobot& n2) {
    if (n1.x != n2.x) return n1.x < n2.x;
    if (n1.y != n2.y) return n1.y < n2.y;
    if (n1.z != n2.z) return n1.z < n2.z;
    return n1.r < n2.r;
}

bool operator==(const nanobot& n1, const nanobot& n2) {
    return n1.x == n2.x && n1.y == n2.y && n1.z == n2.z && n1.r == n2.r;
}

ostream& operator<<(ostream& os, nanobot& n) {
    os << "<" << n.x << "," << n.y << "," << n.z << ">, r=" << n.r;
    return os;
}

inline bool intersect(const nanobot &n1, const nanobot &n2) {
    return dist(n1, n2) <= n1.r + n2.r;
}

inline bool intersect(const set<nanobot> &nc, const nanobot &n2) {
    for (const nanobot& n1: nc) {
        if (!intersect(n1, n2)) {
            return false;
        }
    }
    return true;
}

int main() {
    string line;
    int x, y, z, r;
    vector<nanobot> nanobots;
    while (getline(cin, line)) {
        sscanf(line.c_str(), "pos=<%d,%d,%d>, r=%d", &x, &y, &z, &r);
        nanobot n(x, y, z, r);
        nanobots.push_back(n);
    }
    
    vector<set<nanobot>> nanobot_collections;
    for (const nanobot &n : nanobots) {
        set<nanobot> s {n};
        nanobot_collections.push_back(s);
    }
    
    int max_size = 1;
    int max_i = -1;
    for (int i = 0; i < nanobot_collections.size(); ++i) {
        set<nanobot> &nc = nanobot_collections[i];
        for (const nanobot &n : nanobots) {
            if (nc.count(n) == 0 && intersect(nc, n)) {
                nc.insert(n);
                int curr_size = nc.size();
                if (curr_size > max_size) {
                    max_size = curr_size;
                    max_i = i;
                }
            }
        }
    }
    printf("size of max intersection: %d\n", max_size);
    
    /**
        Ovaj postupak pretpostavlja da je tražena točka jedan od vrhova manhattan sfera.
        To ne mora biti slučaj (npr. ako pričamo o 2D: sjecište dva pravokutnika; pravokutnik je u općem slučaju sjecište dva kvadrata)
    */
    int max_contained_cnt = 0;
    int deltas[] = {-1, 0, 1};
    for (const auto &n: nanobot_collections[max_i]) {
        for (const auto &dx: deltas) {
            for (const auto &dy: deltas) {
                for (const auto &dz: deltas) {
                    bool contained_in_all = true;
                    point p(n.x + dx*n.r, n.y + dy*n.r, n.z + dz*n.r);
                    int contained_cnt = 0;
                    for (const auto &other_n : nanobot_collections[max_i]) {                            
                        if (other_n.contains(p)) {
                            ++contained_cnt;
                            // contained_in_all = false;
                            // break;
                        }
                    }
                    
                    max_contained_cnt = max(contained_cnt, max_contained_cnt);
                    // if (contained_in_all) {
                        // printf("<%d,%d,%d> -> %d\n", p.x, p.y, p.z, contained_cnt);
                    // }
                }
            }
        }
    }
    printf("max_contained_cnt: %d\n", max_contained_cnt);
}